lib/web_ui/lib/src/engine/text/layout_service.dart

```diff
@@ -0,0 +1,328 @@
+            currentLine.getAdditionalWidthTo(nextBreak);
+
+        // For the purpose of max intrinsic width, we don't care if the line
+        // fits within the constraints or not. So we always extend it.
+        if (maxIntrinsicLine.end != nextBreak) {
+          maxIntrinsicLine.extendTo(nextBreak);
+        }
+
+        if (currentLine.width + additionalWidth <= constraints.width) {
+          // TODO(mdebbar): Handle the case when `nextBreak` is just a span end
+          //                that shouldn't extend the line yet.
+
+          // The line can extend to `nextBreak` without overflowing.
+          currentLine.extendTo(nextBreak);
+        } else {
+          // The chunk of text can't fit into the current line.
+          final bool isLastLine =
+              (hasEllipsis && unlimitedLines) || lines.length + 1 == maxLines;
+          if (isLastLine && hasEllipsis) {
+            // We've reached the line that requires an ellipsis to be appended
+            // to it.
+
+            // TODO(mdebbar): Remove this line and implement overflow ellipsis.
+            currentLine.extendTo(nextBreak);
+          } else if (currentLine.isEmpty) {
+            // The current line is still empty, which means we are dealing
+            // with a single block of text that doesn't fit in a single line.
+            // We need to force-break it.
+
+            // TODO(mdebbar): Remove this line and implement force-breaking.
+            currentLine.extendTo(nextBreak);
+          } else {
+            // Normal line break.
+            lines.add(currentLine.build());
+            currentLine = currentLine.nextLine();
+          }
+        }
+      } else {
+        throw UnimplementedError('Unknown span type: ${span.runtimeType}');
+      }
+
+      // ************************************************ //
+      // *** LONGEST LINE && MAX/MIN INTRINSIC WIDTHS *** //
+      // ************************************************ //
+
+      if (longestLine < currentLine.width) {
+        longestLine = currentLine.width;
+      }
+
+      if (minIntrinsicWidth < currentLine.widthOfLastExtension) {
+        minIntrinsicWidth = currentLine.widthOfLastExtension;
+      }
+
+      if (maxIntrinsicLine.end.isHard) {
+        // Max intrinsic width includes the width of trailing spaces.
+        if (maxIntrinsicWidth < maxIntrinsicLine.widthIncludingSpace) {
+          maxIntrinsicWidth = maxIntrinsicLine.widthIncludingSpace;
+        }
+        maxIntrinsicLine = maxIntrinsicLine.nextLine();
+      }
+
+      // ********************************************* //
+      // *** ADVANCE TO THE NEXT SPAN IF NECESSARY *** //
+      // ********************************************* //
+
+      // Only go to the next span if we've reached the end of this span.
+      if (currentLine.end.index >= span.end && spanIndex < spanCount - 1) {
+        span = paragraph.spans[++spanIndex];
+      }
+    }
+  }
+}
+
+class LineBuilder {
```

Comment:
Please add doc comment. Specifically explain usage. Calling LineBuilder.first() and then calling nextLine() as you iterate through spans.

Reply:
Done